# Notebook

In the Model learning step, the prepared dataset from [2_EDA](https://github.com/Rudinius/Bike_usage_Bremen/blob/57e21c8dd687aadc1498f82241cf662840c8b871/2_EDA.ipynb) is loaded. Then different machine learning algorithms are trained and compared to each other.

<a name="content"></a>
# Content

* [1. Import libraries and mount drive](#1)
* [2. Import datasets](#2)
* [3. Transform columns](#3)
* [4. Establish baseline benchmark](#4)
* [5. Training machine learning algorithms](#5)
    * [5.2. XGBoost](#5.2.)
    * [5.3. Multilayer perceptron](#5.3.)
    * [5.4. Recurrent Neural Network](#5.4.)

<a name="1"></a>
# 1.&nbsp;Import libraries
[Content](#content)

In [76]:
# Import libraries
import numpy as np
import pandas as pd
import plotly.express as px
import plotly.graph_objs as go
from plotly.subplots import make_subplots
import random
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import cross_val_score, GridSearchCV, train_test_split, TimeSeriesSplit
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import StandardScaler
import xgboost as xgb
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Input
from keras.layers import Dropout
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.metrics import RootMeanSquaredError

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [3]:
# Install package pyjanitor since it is not part of the standard packages
# of Google Colab

import importlib

# Check if package is installed
package_name = "pyjanitor"
spec = importlib.util.find_spec(package_name)
if spec is None:
    # Package is not installed, install it via pip
    !pip install pyjanitor
else:
    print(f"{package_name} is already installed")

import janitor

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 158.4/158.4 kB 4.0 MB/s eta 0:00:00


/usr/local/lib/python3.10/dist-packages/lazy_loader/__init__.py:185: RuntimeWarning: subpackages can technically be lazily loaded, but it causes the package to be eagerly loaded even if it is already lazily loaded.So, you probably shouldn't use subpackages with this lazy feature.
  warnings.warn(msg, RuntimeWarning)


<a name="2"></a>
#2.&nbsp;Import dataset
[Content](#content)

Next, we will import the processed dataset from [2_EDA](../Bike_usage_Bremen/2_EDA.ipynb).

In [4]:
# Set base url
url = "https://raw.githubusercontent.com/Rudinius/Bike_usage_Bremen/main/data/"

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [5]:
# Import dataset

# We will also parse the date column as datetime64 and set it to the index column
df = pd.read_csv(url + "03_training_data/" + "2023-04-21_df_full.csv",
                         parse_dates=[0], index_col=[0])

# Check the correct loading of dataset
df.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


weekday  graf_moltke_straße_ostseite  \
date                                               
2013-01-01        1                        261.0   
2013-01-02        2                        750.0   
2013-01-03        3                        931.0   
2013-01-04        4                        500.0   
2013-01-05        5                       1013.0   

            graf_moltke_straße_westseite  hastedter_bruckenstraße  \
date                                                                
2013-01-01                         290.0                    381.0   
2013-01-02                         876.0                   1109.0   
2013-01-03                        1015.0                   1603.0   
2013-01-04                         587.0                   1284.0   
2013-01-05                        1011.0                   1284.0   

            langemarckstraße_ostseite  langemarckstraße_westseite  osterdeich  \
date                                                                            
2013-01-01                      312.0                       308.0       870.0   
2013-01-02                     1258.0                      1120.0      2169.0   
2013-01-03                     1556.0                      1480.0      2295.0   
2013-01-04                      703.0                       626.0      1640.0   
2013-01-05                     1856.0                      1621.0      4128.0   

            radweg_kleine_weser  schwachhauser_ring  \
date                                                  
2013-01-01                410.0               391.0   
2013-01-02               1762.0               829.0   
2013-01-03               2287.0              1196.0   
2013-01-04               1548.0              1418.0   
2013-01-05               4256.0              3075.0   

            wachmannstraße_auswarts_sud  ...  tmax  prcp  snow   wdir  wspd  \
date                                     ...                                  
2013-01-01                        514.0  ...   9.1   6.9   0.0  233.0  19.4   
2013-01-02                       1786.0  ...   7.1   1.8   0.0  246.0  20.2   
2013-01-03                       2412.0  ...  10.6   0.9   0.0  257.0  23.8   
2013-01-04                        964.0  ...   9.7   0.0   0.0  276.0  25.2   
2013-01-05                       2065.0  ...   8.6   0.1   0.0  293.0  20.2   

            wpgt    pres  tsun  holiday          vacation  
date                                                       
2013-01-01  50.4  1001.8     0  Neujahr  Weihnachtsferien  
2013-01-02  40.0  1017.5    30      NaN  Weihnachtsferien  
2013-01-03  45.7  1024.5     0      NaN  Weihnachtsferien  
2013-01-04  48.2  1029.5     0      NaN  Weihnachtsferien  
2013-01-05  41.0  1029.9     0      NaN  Weihnachtsferien  

[5 rows x 26 columns]

<a name="3"></a>
# 3. Transform columns
[Content](#content)

We need to transform the columns `Holiday` and `Vacation` using `One-Hot-Encoding` to change the categorical columns to numerical columns. Then we need to drop the original columns.

In [6]:
# Use One Hot Encoder
OH_encoder = OneHotEncoder()
transformed_array = OH_encoder.fit_transform(df.loc[:,"holiday": "vacation"]).toarray()
df_holiday_vacation_transformed = pd.DataFrame(transformed_array,
                              columns=OH_encoder.get_feature_names_out(),
                              index = df.index)
# Drop the columns with Holiday_nan and Vacation_nan as those hold no additional value
df_holiday_vacation_transformed = df_holiday_vacation_transformed.drop(["holiday_nan", "vacation_nan"], axis=1)

# Drop the old categorical columns Holiday and Vacation
df_transformed = df.drop(["holiday", "vacation"], axis=1)

# Add the new columns from OHE
df_transformed = pd.concat([df_transformed, df_holiday_vacation_transformed], axis=1)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


We will add the year, month and day as seperate columns to give the algorithm the chance to pick up more granular and seasonal patterns.

In [28]:
df_date = pd.DataFrame(data = {
    "year": df.index.year,
    "month": df.index.month,
    "day": df.index.day
}, index=pd.to_datetime(df.index.values))

df_transformed_date = (pd.concat([df_date, df_transformed], axis=1)
                        .clean_names(strip_underscores="both"))

# Check dataframe
df_transformed_date.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


year  month  day  weekday  graf_moltke_straße_ostseite  \
2013-01-01  2013      1    1        1                        261.0   
2013-01-02  2013      1    2        2                        750.0   
2013-01-03  2013      1    3        3                        931.0   
2013-01-04  2013      1    4        4                        500.0   
2013-01-05  2013      1    5        5                       1013.0   

            graf_moltke_straße_westseite  hastedter_bruckenstraße  \
2013-01-01                         290.0                    381.0   
2013-01-02                         876.0                   1109.0   
2013-01-03                        1015.0                   1603.0   
2013-01-04                         587.0                   1284.0   
2013-01-05                        1011.0                   1284.0   

            langemarckstraße_ostseite  langemarckstraße_westseite  osterdeich  \
2013-01-01                      312.0                       308.0       870.0   
2013-01-02                     1258.0                      1120.0      2169.0   
2013-01-03                     1556.0                      1480.0      2295.0   
2013-01-04                      703.0                       626.0      1640.0   
2013-01-05                     1856.0                      1621.0      4128.0   

            ...  holiday_pfingstmontag  holiday_reformationstag  \
2013-01-01  ...                    0.0                      0.0   
2013-01-02  ...                    0.0                      0.0   
2013-01-03  ...                    0.0                      0.0   
2013-01-04  ...                    0.0                      0.0   
2013-01-05  ...                    0.0                      0.0   

            holiday_tag_der_arbeit  holiday_tag_der_deutschen_einheit  \
2013-01-01                     0.0                                0.0   
2013-01-02                     0.0                                0.0   
2013-01-03                     0.0                                0.0   
2013-01-04                     0.0                                0.0   
2013-01-05                     0.0                                0.0   

            vacation_herbstferien  vacation_osterferien  \
2013-01-01                    0.0                   0.0   
2013-01-02                    0.0                   0.0   
2013-01-03                    0.0                   0.0   
2013-01-04                    0.0                   0.0   
2013-01-05                    0.0                   0.0   

            vacation_pfingstferien  vacation_sommerferien  \
2013-01-01                     0.0                    0.0   
2013-01-02                     0.0                    0.0   
2013-01-03                     0.0                    0.0   
2013-01-04                     0.0                    0.0   
2013-01-05                     0.0                    0.0   

            vacation_weihnachtsferien  vacation_winterferien  
2013-01-01                        1.0                    0.0  
2013-01-02                        1.0                    0.0  
2013-01-03                        1.0                    0.0  
2013-01-04                        1.0                    0.0  
2013-01-05                        1.0                    0.0  

[5 rows x 43 columns]

Now, after all those transformations, we have out final dataset, to train our machine learning algorithms on.

<a name="4"></a>
# 4. Establish baseline benchmark
[Content](#content)


For our current task of creating model a to predict the amount of cyclers for a given day, we do not have any baseline metric score to measure our model against.
For this reason, we will create a naive baseline model. For this, we will simply predict the amount of a day based on the value of previous day.

In [8]:
# Evaluate the model's performance using RMSE

# Select the `Total` column as our y_dev and preds arrays
y_dev = preds = df.loc[:,"total"]

rmse = 0
length = y_dev.shape[0]

# Loop from 0 to second last entry, as we can only use seconds last entry to
# predict the last entry of series
for i in range(length-1):
    # The mean_sqared_error function expects an array as input, therfore we
    # concatenate the range from current value to current value + 1 (excluding)
    rmse += np.sqrt(mean_squared_error(y_dev[i+1:i+2], preds[i:i+1]))

# Divide rmse value by number of pairs
rmse = rmse / (length-1)
print("RMSE: %f" % (rmse))

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


RMSE: 9083.225418


If we were naivly predicting the current value with the last value, we get an error over the entire dataset of approximately $9,100$.

This is our naive benchmark to compare our model against.

Another method would be to predict the value of a given day by the average of all the other equal days in the dataset(e.g. to predict 18.08.2017, we take the average of all other 18.08. days in the dataset).

In [9]:
# Initialize squared error
se = 0

# Get the total number of examples
m = df.shape[0]

for i in df.index:
    day = i.day
    month = i.month
    year = i.year

    # create a mask for given day but exclude the day we want to predict
    mask = (df.index.day == day) & (df.index.month == month) & (df.index.year != year)

    # Get value for current day and mean values of all the other same days in the dataset
    y_true = df.loc[i, "total"]
    y_pred = df.loc[mask,"total"].mean()

    # Calculate the squared error
    se += (y_true - y_pred)**2

# Calculate mean squared error
mse = se / m

# Calcualte root mean squared error
rmse = np.sqrt(mse)

print(f"RMSE: {rmse}")

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


RMSE: 10282.852522084184


With this second approach, of average all our previous values for the given day and using this as our forecast, we get an error over the entire dataset of approximately $10,300$.

The error of this second naive approach is close to the first approach.
Both approaches could be seen as human-level as this would be a typical approach of a human, to predict the value of any given day. A domain expert, who also looks at more data and e.g., compares also the temperatures, could come up with better estimates. However humans are typically not very good in accurately predicting complex time-series data. The expected Bayes error (least possible error) should therefore be much lower.

<a name="5"></a>
# 5. Training machine learning algorithms
[Content](#content)

We are going to train 1 shallow machine learning algorithm and 2 deep machine learning algorithms to be able to compare performances. Those are:

* XGBoost
* Multilayer Perceptron (MLP -- standard NN)
* Recurrent Neural Network (RNN)

DEMO

In contrast to RNNs where the algorithm takes automically the datapoints of previous timesteps into account, XGBoost and MLPs do not have direct access to the sequential data of previous time steps (lag observation). Those algorithms have only indirect knowledge via the learned model parameters. RNNs however directly include the previous timestep for learning the parameters of the current timestep.

For this reason, we will prepare our training data for XGBoost and MLP in sequences of length `n`. This means we will take the last `n` observations (datapoints), to forecast the `n+1` observation (e.g., datapoints of day 1, 2, and 3 are used to predict the value of day 4).

First, we create a function, to split our data into those mentioned sequences.

In [23]:
# Split a multivariate sequence into samples
# This function assumes, that the target is the last element
def split_sequences(sequences, n_steps):
    X, y = list(), list()
    for i in range(len(sequences)):
        # Find the end of this pattern
        end_ix = i + n_steps

        # Check if we are beyond the dataset
        if end_ix > len(sequences):
            break

        # Gather input and output parts of the pattern
        seq_x, seq_y = sequences.iloc[i:end_ix, :-1], sequences.iloc[end_ix-1, -1]
        X.append(seq_x)
        y.append(seq_y)

    # Return the arrays with the splitted sequences
    return np.array(X), np.array(y)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Rearange the features, so that `total` is at the end as needed for the funciton above.
# Throw out all not wanted features and targets.

# Higly correlated features have been removed (tavg, tmin, wpgt)
# `wdir` has been removed due to no correlation

features = ['year', 'month', 'day', 'weekday', 'tmax', 'prcp',
            'snow', 'wspd', 'pres', 'tsun',
            'holiday_1_weihnachtsfeiertag', 'holiday_2_weihnachtsfeiertag',
            'holiday_christi_himmelfahrt', 'holiday_karfreitag', 'holiday_neujahr',
            'holiday_ostermontag', 'holiday_pfingstmontag', 'holiday_reformationstag',
            'holiday_tag_der_arbeit', 'holiday_tag_der_deutschen_einheit',
            'vacation_herbstferien', 'vacation_osterferien', 'vacation_pfingstferien',
            'vacation_sommerferien', 'vacation_weihnachtsferien', 'vacation_winterferien',
            'total']

# Load the data into a pandas dataframe
data = df_transformed_date[features]
data

In [53]:
# Split the training data into sequences with 4 days each
# 3 previous days and 1 day to forecast
data_X, data_y = split_sequences(data, 4)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [54]:
print(f"Shape train_X: {data_X.shape}, shape train_y: {data_y.shape}")

Shape train_X: (3649, 4, 26), shape train_y: (3649,)


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


`train_X` is of shape (3649, 4, 26). The first dimension is the number of sequences, in this case 3649. Since we added some days together to create a sequence, the total amount of sequences compared to the previous number of days declined slighty. The second dimension is the number of time steps per sample, in this case 4. This is the number of days per sequence and the value specified to the function above. The last dimension is the number of our features used to predict the value like holidays, temperature and so on.

`train_y` is an array of lenght 3649 and matches therefore the number of sequences in `train_X`.

<a name="5.1."></a>
## 5.1. Adding sequential data to our model

[Content](#content)

In contrast to RNNs where the algorithm takes automically the datapoints of previous timesteps into account, XGBoost and MLPs do not have direct access to the sequential data of previous time steps.
Those algorithms have only indirect knowledge via the learned model parameters. RNNs however directly include the previous timestep for learning the parameters of the current timestep.

We will add the data points of the previous time steps as features to the feature vector.

In this case, we will only add the last 3 values, as the observed improvement of accuracy (RMSE score) is drastically decressing with each further time step added after 3 steps.

Improvements:
* 1 day 6291 2,8%
* 2 day 6157 2,1%
* 3 day 6120 0,6%

The following code creates a dataframe with a variable amount of time:

In [ ]:
# Create empty new dataframe
df_lagged_days = pd.DataFrame({})

# Select the number of lagged days
go_back_x_days = 3

for i in range(go_back_x_days):
    # Shift the values in "Total" by i and assign to new column prev_Total_i+1
    df_lagged_days[f'prev_total_{i+1}'] = df_transformed_date['total'].shift(i+1)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


year  month  day  weekday  graf_moltke_straße_ostseite  \
2013-01-01  2013      1    1        1                        261.0   
2013-01-02  2013      1    2        2                        750.0   
2013-01-03  2013      1    3        3                        931.0   
2013-01-04  2013      1    4        4                        500.0   
2013-01-05  2013      1    5        5                       1013.0   
...          ...    ...  ...      ...                          ...   
2022-12-27  2022     12   27        1                        693.0   
2022-12-28  2022     12   28        2                        643.0   
2022-12-29  2022     12   29        3                        654.0   
2022-12-30  2022     12   30        4                        757.0   
2022-12-31  2022     12   31        5                        371.0   

            graf_moltke_straße_westseite  hastedter_bruckenstraße  \
2013-01-01                         290.0                    381.0   
2013-01-02                         876.0                   1109.0   
2013-01-03                        1015.0                   1603.0   
2013-01-04                         587.0                   1284.0   
2013-01-05                        1011.0                   1284.0   
...                                  ...                      ...   
2022-12-27                         612.0                   1495.0   
2022-12-28                         585.0                   1076.0   
2022-12-29                         648.0                   1076.0   
2022-12-30                         665.0                   1076.0   
2022-12-31                         367.0                    557.0   

            langemarckstraße_ostseite  langemarckstraße_westseite  osterdeich  \
2013-01-01                      312.0                       308.0       870.0   
2013-01-02                     1258.0                      1120.0      2169.0   
2013-01-03                     1556.0                      1480.0      2295.0   
2013-01-04                      703.0                       626.0      1640.0   
2013-01-05                     1856.0                      1621.0      4128.0   
...                               ...                         ...         ...   
2022-12-27                     1062.0                       915.0      2123.0   
2022-12-28                      884.0                       820.0      1819.0   
2022-12-29                     1014.0                       907.0      2013.0   
2022-12-30                     1106.0                       976.0      2088.0   
2022-12-31                      656.0                       404.0      1156.0   

            ...  holiday_tag_der_deutschen_einheit  vacation_herbstferien  \
2013-01-01  ...                                0.0                    0.0   
2013-01-02  ...                                0.0                    0.0   
2013-01-03  ...                                0.0                    0.0   
2013-01-04  ...                                0.0                    0.0   
2013-01-05  ...                                0.0                    0.0   
...         ...                                ...                    ...   
2022-12-27  ...                                0.0                    0.0   
2022-12-28  ...                                0.0                    0.0   
2022-12-29  ...                                0.0                    0.0   
2022-12-30  ...                                0.0                    0.0   
2022-12-31  ...                                0.0                    0.0   

            vacation_osterferien  vacation_pfingstferien  \
2013-01-01                   0.0                     0.0   
2013-01-02                   0.0                     0.0   
2013-01-03                   0.0                     0.0   
2013-01-04                   0.0                     0.0   
2013-01-05                   0.0                     0.0   
...                          ...                     ...   
2022-12-27              

<a name="5.2."></a>
## 5.2. XGBoost
[Content](#content)

For XGBoost, we will add the dataframe `df_lagged_days` to our dataset.

In [ ]:
# Concat new dataframe with old dataframe
# Using bfill strategy on dataset since the first few days will have NaN values
# Using pyjanitor to clean up names
df_transformed_date_lagged = (pd.concat([df_transformed_date, df_lagged_days], axis=1)
                              .fillna(method="bfill")
                              .clean_names(strip_underscores="both"))

# Check output
df_transformed_date_lagged

In [ ]:
# Load the data into a pandas dataframe
data = df_transformed_date_lagged
# Define the features and target
# Higly correlated features have been removed (tavg, tmin, wpgt)
# `wdir` has been removed due to no correlation

#features = df_transformed_date_lagged.columns ??
#features.drop(["tavg", "tmin", "wpgt", "wdir", "total"], axis = 1)

features = ['year', 'month', 'day', 'weekday', 'tmax', 'prcp',
            'snow', 'wspd', 'pres', 'tsun',
            'holiday_1_weihnachtsfeiertag', 'holiday_2_weihnachtsfeiertag',
            'holiday_christi_himmelfahrt', 'holiday_karfreitag', 'holiday_neujahr',
            'holiday_ostermontag', 'holiday_pfingstmontag', 'holiday_reformationstag',
            'holiday_tag_der_arbeit', 'holiday_tag_der_deutschen_einheit',
            'vacation_herbstferien', 'vacation_osterferien', 'vacation_pfingstferien',
            'vacation_sommerferien', 'vacation_weihnachtsferien', 'vacation_winterferien',
            'prev_total_1', 'prev_total_2', 'prev_total_3']

target = 'total'

# Split the data into training and dev sets
#X_train, x_dev, y_train, y_dev = train_test_split(data[features], data[target],
#                                                    test_size=0.2, shuffle=True, random_state=0)



/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [60]:
# USE THE TEST VERSION WITH SEQUENCES AND FLATTENED
X_train, X_dev, y_train, y_dev = train_test_split(data_X, data_y,
                                                    test_size=0.2, shuffle=False, random_state=0)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


# Flatten

To feed those sequences into out models, we need to flatten them.

In [57]:
X_train[0]

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


array([[2.0130e+03, 1.0000e+00, 1.0000e+00, 1.0000e+00, 9.1000e+00,
        6.9000e+00, 0.0000e+00, 1.9400e+01, 1.0018e+03, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 1.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 1.0000e+00,
        0.0000e+00],
       [2.0130e+03, 1.0000e+00, 2.0000e+00, 2.0000e+00, 7.1000e+00,
        1.8000e+00, 0.0000e+00, 2.0200e+01, 1.0175e+03, 3.0000e+01,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 1.0000e+00,
        0.0000e+00],
       [2.0130e+03, 1.0000e+00, 3.0000e+00, 3.0000e+00, 1.0600e+01,
        9.0000e-01, 0.0000e+00, 2.3800e+01, 1.0245e+03, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00,
      

In [69]:
print(X_train.shape)
X_train_flattened = X_train.reshape(X_train.shape[0], -1)
X_dev_flattened = X_dev.reshape(X_dev.shape[0], -1)
X_train_flattened.shape

(2919, 4, 26)


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


(2919, 104)

Finally, we will standardize our dataset. Standarization will generally improve learning speed of the models and can help to improve the accuarcy of the model.

In [71]:
# Standardize and fit to the training set only
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train_flattened)

# Apply the same standardization to the dev set
x_dev_scaled = scaler.transform(X_dev_flattened)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Using `GridSearchCV` to select optimal parameters.

In [ ]:
params = {
    'learning_rate': [0.01, 0.05, 0.1],
    'max_depth': [3, 5, 7],
    'n_estimators': [700, 1000, 1200],
    'subsample': [0.5, 0.8, 1.0],
    'colsample_bytree': [0.5, 0.8, 1.0],
    'reg_alpha': [0.0, 0.5, 1.0],
    'reg_lambda': [0.8, 1.0, 2.0],
    'min_child_weight': [1, 5, 10]
}

xg_reg = xgb.XGBRegressor(objective='reg:squarederror')

grid_search = GridSearchCV(xg_reg, param_grid=params, cv=5, n_jobs=-1, verbose=2)

grid_search.fit(X_train, y_train)

print(grid_search.best_params_)

In [93]:
%%time
### SCALED ###
# Build the XGBoost regressor model with selected hyper parameters
xg_reg = xgb.XGBRegressor(objective ='reg:squarederror', colsample_bytree = 1.0, learning_rate = 0.01,
                          max_depth = 12, n_estimators = 1000, reg_alpha = 0.5, reg_lambda = 5.0,
                          subsample = 1.0, min_child_weight=5)
#print(X_train_scaled.shape)
#print(y_train.shape)
xg_reg.fit(X_train_scaled, y_train)



# Evaluate the model's performance using RMSE
# Predict on the train set
y_preds_train = xg_reg.predict(X_train_scaled)
print(y_preds_train.shape)
print(y_preds_train)

# Training set
rmse_train = np.sqrt(mean_squared_error(y_train, y_preds_train))
print("Train RMSE: %f" % (rmse_train))

# Predict on the dev set
y_dev_preds = xg_reg.predict(x_dev_scaled)
# Dev set
rmse_dev = np.sqrt(mean_squared_error(y_dev, y_dev_preds))
print("Dev RMSE: %f" % (rmse_dev))

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


(2919,)
[13574.302 30045.324 25157.857 ... 17235.213 17710.488 13389.212]
Train RMSE: 909.820455
Dev RMSE: 7415.265703
CPU times: user 1min 6s, sys: 377 ms, total: 1min 7s
Wall time: 38.1 s


In [ ]:
%%time
# Build the XGBoost regressor model with selected hyper parameters
xg_reg = xgb.XGBRegressor(objective ='reg:squarederror', colsample_bytree = 1.0, learning_rate = 0.01,
                          max_depth = 12, n_estimators = 1000, reg_alpha = 0.5, reg_lambda = 5.0,
                          subsample = 1.0, min_child_weight=5)

xg_reg.fit(X_train, y_train)



# Evaluate the model's performance using RMSE
# Predict on the train set
preds_train = xg_reg.predict(X_train)
print(preds_train)
print(preds_train.shape)

# Training set
rmse_train = np.sqrt(mean_squared_error(y_train, preds_train))
print("Train RMSE: %f" % (rmse_train))

# Predict on the dev set
preds_dev = xg_reg.predict(x_dev)

# Dev set
rmse_dev = np.sqrt(mean_squared_error(y_dev, preds_dev))
print("Dev RMSE: %f" % (rmse_dev))

/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


[15197.027 33466.727 15071.921 ... 28035.055 32195.223 57779.17 ]
(2921,)
Train RMSE: 1216.824632
Test RMSE: 6054.719014
CPU times: user 37.8 s, sys: 490 ms, total: 38.3 s
Wall time: 28.2 s


* 02.04.23: 6468
* 10.04.23: 6120

As an alternative, instead of `train_test_split`, we try `TimeSeriesSplit` and compare the performance.

In [ ]:
# Split the data into training and dev sets
tscv = TimeSeriesSplit(n_splits=5)
for train_index, dev_index in tscv.split(data):
    X_train, x_dev = data.iloc[train_index][features], data.iloc[dev_index][features]
    y_train, y_dev = data.iloc[train_index][target], data.iloc[dev_index][target]

    xg_reg = xgb.XGBRegressor(objective='reg:squarederror', colsample_bytree=1.0, learning_rate=0.01,
                              max_depth=12, n_estimators=1000, reg_alpha=0.1, reg_lambda=1.0,
                              subsample=0.8, min_child_weight=5)

    xg_reg.fit(X_train, y_train)

    # Predict on the dev set
    preds = xg_reg.predict(x_dev)

    # Evaluate the model's performance using RMSE
    rmse = np.sqrt(mean_squared_error(y_dev, preds))
    print("RMSE: %f" % (rmse))


RMSE: 7127.157045
RMSE: 6843.915770
RMSE: 6808.748428
RMSE: 7121.689867
RMSE: 6898.040513


`train_test_split` has a better performance on the accuracy of the model than `TimeSeriesSplit`.

Plot random different data points to analyze big differences between prediction and actual value in more detail.

In [ ]:
# Select 100 random datapoints from y_dev
random_indices = random.sample(range(len(y_dev)), 100)
y_dev_sample = y_dev[random_indices]
preds_sample = preds[random_indices]

# Create plot using plotly express
fig = px.scatter()
fig.add_scatter(x=y_dev_sample.index, y=y_dev_sample, mode='markers', name='y_dev', marker=dict(color='blue'))
fig.add_scatter(x=y_dev_sample.index, y=preds_sample, mode='markers', name='preds', marker=dict(color='red'))
fig.update_layout(title='y_dev vs. preds, random values', xaxis_title='Date', yaxis_title='Values')
fig.show()

/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


<a name="5.3."></a>
## 5.3. Multilayer Perceptron
[Content](#content)

TO CHECK: Performance when standardizing/normalizing dataset
TO CHECK: When adding the prev_total values, the accuracy becomes less. Check later with optimized MLP again

In [84]:
%%time
from tensorflow.keras import regularizers
from keras.regularizers import l2

regl2 = 20
dropout = 0.0 # 0.2 best

# Define the architecture of the MLP with L2 regularization
model = Sequential()
model.add(Input(shape=(104,)))
#model.add(Input(shape=(29,)))
model.add(Dense(units=128, activation='relu', kernel_regularizer=regularizers.L1L2(l1=1e5, l2=1e5)))
model.add(Dropout(dropout))
model.add(Dense(units=64, activation='relu', kernel_regularizer=regularizers.L1L2(l1=1e5, l2=1e5)))
model.add(Dropout(dropout))
model.add(Dense(units=32, activation='relu', kernel_regularizer=regularizers.L1L2(l1=1e5, l2=1e5)))
model.add(Dropout(dropout))
#model.add(Dense(units=512, activation='relu', kernel_regularizer=l2(regl2)))
#model.add(Dense(units=512, activation='relu', kernel_regularizer=l2(regl2)))
#model.add(Dense(units=512, activation='relu', kernel_regularizer=l2(regl2)))
#model.add(Dense(units=512, activation='relu', kernel_regularizer=l2(regl2)))
#model.add(Dense(units=512, activation='relu', kernel_regularizer=l2(regl2)))
#model.add(Dense(units=512, activation='relu', kernel_regularizer=l2(regl2)))
#model.add(Dense(units=512, activation='relu', kernel_regularizer=l2(regl2)))
#model.add(Dense(units=512, activation='relu', kernel_regularizer=l2(regl2)))
model.add(Dense(units=1, activation='linear', kernel_regularizer=l2(regl2)))  # Output layer with 1 neuron for regression

#print(model.summary())

# Compile the model with mean squared error (MSE) loss, and root mean square error (RMSE) as metric
# Use Adam optimizer with learning rate
optimizer = Adam(learning_rate=0.1)
model.compile(optimizer=optimizer, loss='mse', metrics=[RootMeanSquaredError()])

# Train the model and save the learning history, use x_dev and y_dev for validation
history = model.fit(X_train_scaled, y_train, epochs=500, batch_size=256, validation_data=(x_dev_scaled, y_dev))

Epoch 1/500


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


12/12 [==============================] - 1s 32ms/step - loss: 1271213824.0000 - root_mean_squared_error: 34101.7695 - val_loss: 1226347008.0000 - val_root_mean_squared_error: 34380.6953
Epoch 2/500
12/12 [==============================] - 0s 12ms/step - loss: 1054186560.0000 - root_mean_squared_error: 31174.6445 - val_loss: 545760896.0000 - val_root_mean_squared_error: 17953.6230
Epoch 3/500
12/12 [==============================] - 0s 16ms/step - loss: 402880800.0000 - root_mean_squared_error: 13608.0996 - val_loss: 323133024.0000 - val_root_mean_squared_error: 10437.0889
Epoch 4/500
12/12 [==============================] - 0s 12ms/step - loss: 294749312.0000 - root_mean_squared_error: 9460.1338 - val_loss: 259832080.0000 - val_root_mean_squared_error: 9534.3770
Epoch 5/500
12/12 [==============================] - 0s 13ms/step - loss: 232685536.0000 - root_mean_squared_error: 8545.1729 - val_loss: 228024336.0000 - val_root_mean_squared_error: 9633.5449
Epoch 6/500
12/12 [==============

With L2 0.0

with L2 5.0

In [ ]:
history_reg0 = history

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
history_reg1000 = history

/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
history_reg0_not_scaled = history

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
model.layers[0].get_config()

Epoch 100/100
46/46 [==============================] - 1s 23ms/step - loss: 6617481.5000 - root_mean_squared_error: 2572.4465 - val_loss: 67260184.0000 - val_root_mean_squared_error: 8201.2305
CPU times: user 2min 43s, sys: 6.67 s, total: 2min 50s
Wall time: 2min 23

In [ ]:
hist_train_rmse_reg0 = np.array(history_reg0.history["root_mean_squared_error"])
hist_dev_rmse_reg0 = np.array(history_reg0.history["val_root_mean_squared_error"])
hist_train_rmse_reg0_not_scaled = np.array(history_reg0_not_scaled.history["root_mean_squared_error"])
hist_dev_rmse_reg0_not_scaled = np.array(history_reg0_not_scaled.history["val_root_mean_squared_error"])

# Create the array for the x axis, starting from 1
x = np.arange(1, len(history_reg0.history["root_mean_squared_error"])+1)

# Create a line chart with two lines using Plotly Express
fig = px.line(title='train RMSE vs dev RMSE')
fig.add_scatter(x=x, y=hist_train_rmse_reg0, mode='lines+markers', name='Train L2=0', line=dict(color='red'))
fig.add_scatter(x=x, y=hist_dev_rmse_reg0, mode='lines+markers', name='Dev L2=0', line=dict(color='blue'))
fig.add_scatter(x=x, y=hist_train_rmse_reg0_not_scaled, mode='lines+markers', name='Train L2=0 not scaled', line=dict(color='gray'))
fig.add_scatter(x=x, y=hist_dev_rmse_reg0_not_scaled, mode='lines+markers', name='Dev L2=0 not scaled', line=dict(color='black'))

# Set chart title and axis labels
fig.update_layout(
    xaxis_title='Epochs',
    yaxis_title='RMSE'
)

# Show the chart
fig.show()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
hist_train_rmse = np.array(history.history["root_mean_squared_error"])
hist_dev_rmse = np.array(history.history["val_root_mean_squared_error"])

# Create the array for the x axis, starting from 1
x = np.arange(1, len(history.history["root_mean_squared_error"])+1)

# Create a line chart with two lines using Plotly Express
fig = px.line(title='train RMSE vs dev RMSE')
fig.add_scatter(x=x, y=hist_train_rmse, mode='lines+markers', name='Train', line=dict(color='red'))
fig.add_scatter(x=x, y=hist_dev_rmse, mode='lines+markers', name='Dev', line=dict(color='blue'))

# Set chart title and axis labels
fig.update_layout(
    xaxis_title='Epochs',
    yaxis_title='RMSE'
)

# Show the chart
fig.show()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



In [ ]:
preds_train = model.predict(X_train_scaled)

# Evaluate the model's performance using RMSE
rmse_train = np.sqrt(mean_squared_error(y_train, preds_train))
print("Trai RMSE: %f" % (rmse_train))

preds_dev = model.predict(x_dev_scaled)

# Evaluate the model's performance using RMSE
rmse_dev = np.sqrt(mean_squared_error(y_dev, preds_dev))
print("Dev RMSE: %f" % (rmse_dev))

/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



92/92 [==============================] - 1s 10ms/step
Trai RMSE: 2623.097985
23/23 [==============================] - 0s 8ms/step
Test RMSE: 8387.871121


<a name="5.4."></a>
## 5.4. Recurrent Neural Network
[Content](#content)

In [ ]:
# Load the data into a pandas dataframe
data = df_transformed_date

# Define the features and target
# Higly correlated features have been removed (tavg, tmax, wpgt)
features = ['year', 'month', 'day', 'weekday', 'tmax', 'prcp',
            'snow', 'wspd', 'pres', 'tsun',
            'holiday_1_weihnachtsfeiertag', 'holiday_2_weihnachtsfeiertag',
            'holiday_christi_himmelfahrt', 'holiday_karfreitag', 'holiday_neujahr',
            'holiday_ostermontag', 'holiday_pfingstmontag', 'holiday_reformationstag',
            'holiday_tag_der_arbeit', 'holiday_tag_der_deutschen_einheit',
            'vacation_herbstferien', 'vacation_osterferien', 'vacation_pfingstferien',
            'vacation_sommerferien', 'vacation_weihnachtsferien', 'vacation_winterferien']

target = 'Total'

# Split the data into training and dev sets
X_train, x_dev, y_train, y_dev = train_test_split(data[features], data[target],
                                                    test_size=0.2, shuffle=True, random_state=0)

In [ ]:

import tensorflow as tf
import numpy as np
from keras.models import Sequential
from keras.layers import LSTM, Dense
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split
from keras.optimizers import Adam

# Assuming X_train, x_dev, y_train, y_dev are already defined and contain the appropriate data




num_samples = X_train.shape[0]
# Build the LSTM model
"""
model = Sequential()
model.add(LSTM(units=256, input_shape=(1, X_train.shape[1]), activation='relu'))
# Add a dense output layer with a single unit (for regression) and no activation function
model.add(Dense(units=1))
"""
"""
model = tf.keras.models.Sequential([
    # Shape [batch, time, features] => [batch, time, lstm_units]
    tf.keras.layers.LSTM(64),
    # Shape => [batch, time, features]
    tf.keras.layers.Dense(units=1)
])
"""


# Build the LSTM model
model = tf.keras.models.Sequential([
    tf.keras.layers.LSTM(256, activation='relu'),
    # Shape [batch, time, features] => [batch, time, lstm_units]
    tf.keras.layers.LSTM(128, activation='relu'),
    # Add another LSTM layer with 128 units and ReLU activation
    tf.keras.layers.LSTM(64, activation='relu'),
    # Shape => [batch, time, features]
    tf.keras.layers.Dense(units=1)
])

X_train_3d = np.reshape(X_train.to_numpy(), (num_samples, 1, X_train.shape[1]))  # Fix variable name
model.compile(loss=tf.keras.losses.MeanSquaredError(),
              optimizer=Adam(learning_rate=0.001),
              metrics=[tf.keras.metrics.MeanAbsoluteError()])

# Train the model
model.fit(X_train_3d, y_train, epochs=100, batch_size=64, verbose=1)

Epoch 1/100


/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



ValueError: ignored

In [ ]:
# Predict on the dev set
#x_dev_3d = np.reshape(x_dev.to_numpy(), (x_dev.shape[0], 1, x_dev.shape[1]))  # Reshape x_dev

num_samples = x_dev.shape[0]
x_dev_3d = np.reshape(x_dev.to_numpy(), (num_samples, 1, x_dev.shape[1]))  # Fix variable name
preds = model.predict(x_dev_3d)

# Evaluate the model's performance using RMSE
rmse = np.sqrt(mean_squared_error(y_dev, np.reshape(preds, (y_dev.shape[0]))))
print("RMSE: %f" % (rmse))


23/23 [==============================] - 0s 6ms/step
2018-03-11    26957.0
2014-07-27    37410.0
2016-09-30    39593.0
2022-06-05    50288.0
2017-09-29    41285.0
               ...   
2013-08-06    37690.0
2014-11-02    29683.0
2020-06-27    33080.0
2016-08-31    53149.0
2022-01-16    12165.0
Name: Total, Length: 731, dtype: float64
[30535.756 37923.53  32049.592 36079.492 31525.645 28276.818 31182.781
 28032.14  26946.822 35099.41  40275.06  31049.348 38569.285 31857.996
 30947.023 28566.012 28895.145 29982.74  36054.23  28999.047 29830.938
 30775.758 27452.115 31243.29  27236.404 27673.469 29084.553 30560.8
 30556.438 34519.117 30605.236 31264.998 33066.836 20967.12  37602.754
 32578.305 39709.023 29811.994 28410.668 33195.67  27754.33  31554.383
 27051.059 28399.223 35825.227 27933.12  39415.55  35235.44  35060.434
 32877.6   31883.713 31583.65  28782.07  34868.4   32103.193 28370.572
 26299.674 29966.117 36611.188 40976.992 27086.205 35797.164 28439.223
 33622.242 28705.676 38175.